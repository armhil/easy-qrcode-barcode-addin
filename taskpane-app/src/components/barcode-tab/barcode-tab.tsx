import React, { useState, useRef } from 'react';
import { Button } from '@fluentui/react-components';
import { AddinUtils, LoggingUtils } from 'easy-addins-utils';
import Barcode from 'react-barcode';
import { SliderWrapper } from '../slider';
import { InputWrapper } from '../input';

// Barcode component
export function BarcodeTab() {
  React.useEffect(() => {
      LoggingUtils.Trace('qrbar-barcode'); 
  }, []);

  const ref = useRef(null);
  const [text, setText] = useState("");
  const [width, setWidth] = useState(1);
  const [height, setHeight] = useState(100);

  const WidthSlider = () => <SliderWrapper label="Width slider" max={50} min={0.4} step={0.2} value={width} setValue={setWidth} />
  const HeightSlider = () => <SliderWrapper  label="Height slider" max={150} min={50} step={5} value={height} setValue={setHeight} />
  const updateText = (input: string | undefined) => {
    if (input)
      setText(input);
  };

  const insertImageFromCanvas = () => {
    const canvasValue = getCanvasURL();
    AddinUtils.InsertImage(canvasValue, () => {});
  }

  // Insert image to Word
  const insertImage = () => {
    if (!text) {
        AddinUtils.GetText((t: string) => {
          setText(t);
          insertImageFromCanvas();
        });
    }
    else if (text && text.length) {
      insertImageFromCanvas();
    }
    else {
        //AddinUtils.("Could get the data URL from the canvas");
    }
  };

  const getCanvasURL = () => {
    const dataValue: any = ref?.current?.['children']?.[0];
    if (dataValue) {
      let str = dataValue.toDataURL();
      str = str.split('data:image/png;base64,')[1];
      return str;
    }
  };

  return (
    <>
      <div data-testid='input-barcode-text'>
        <InputWrapper updateText={updateText} value={text} label="Barcode text"/>
      </div>
      <WidthSlider />
      <HeightSlider />
      <div ref={ref} data-testid='canvas'>
        <Barcode displayValue={false} renderer='canvas'
          width={width} height={height} value={text}/>
      </div>
      <div>
          <Button data-testid='btn-insert-image' appearance='primary' onClick={insertImage}>Insert Image</Button>
      </div>
    </>
  );
}
