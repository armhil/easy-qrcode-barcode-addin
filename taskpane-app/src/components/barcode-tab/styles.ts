import { makeStyles } from '@fluentui/react-components';

export const useInputStyles = makeStyles({
	width: {'width': '100%'}
})