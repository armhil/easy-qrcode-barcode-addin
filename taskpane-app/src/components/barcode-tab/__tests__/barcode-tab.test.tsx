import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { AddinUtils, LoggingUtils } from 'easy-addins-utils';
import { BarcodeTab } from '../barcode-tab';

describe('barcode rendering', () => {
  it('should match with snapshot', () => {
    const tree = renderer.create(<BarcodeTab />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should log a trace', () => {
    const traceMockFn = jest.fn();
    LoggingUtils.Trace = traceMockFn;

    render(<BarcodeTab />);
    expect(traceMockFn).toHaveBeenCalledTimes(1);
    expect(traceMockFn).toHaveBeenCalledWith('qrbar-barcode');
  });

  it('should render links with correct attributes', () => {
    const dom = render(<BarcodeTab />);

    const links = dom.queryAllByRole('link');
    // no links on the page
    expect(links.length).toBe(0);
  });

  it('should not render the canvas on empty text', () => {
    const dom = render(<BarcodeTab />);

    const canvas = dom.queryByTestId('barcode-canvas');
    expect(canvas).toBeNull();
  });

  it('should render the canvas correctly on non-empty text', () => {
    const dom = render(<BarcodeTab />);
    // enter value testing
    const input = dom.queryAllByRole('textbox');
    act(() => {
      fireEvent.change(input[0] as HTMLElement, {
        target: { value: 'testing' },
      });
    });
    const canvas = dom.getByTestId('barcode-canvas');
    expect(canvas).not.toBeEmptyDOMElement();
    expect((canvas.firstChild as HTMLElement).tagName).toBe('CANVAS');
  });

  it('should attempt to read from document if textbox is empty', () => {
    AddinUtils.InsertImage = jest.fn();
    AddinUtils.GetText = jest.fn();
    window.Office = {
      context: {
        document: {
          getSelectedDataAsync: jest.fn(),
        },
      },
    };
    const dom = render(<BarcodeTab />);

    const insertButton = dom.queryByRole('button');
    act(() => {
      insertButton!.click();
    });
    expect(AddinUtils.GetText).toHaveBeenCalledTimes(1);
  });

  it('should attempt to insert image if textbox is not empty', () => {
    AddinUtils.InsertImage = jest.fn();
    AddinUtils.GetText = jest.fn();
    const dom = render(<BarcodeTab />);

    const input = dom.queryAllByRole('textbox');
    act(() => {
      fireEvent.change(input[0] as HTMLElement, {
        target: { value: 'testing' },
      });
    });
    const insertButton = dom.queryByRole('button');
    act(() => {
      insertButton!.click();
    });
    expect(AddinUtils.GetText).toHaveBeenCalledTimes(0);
    expect(AddinUtils.InsertImage).toHaveBeenCalledTimes(1);
  });
});
