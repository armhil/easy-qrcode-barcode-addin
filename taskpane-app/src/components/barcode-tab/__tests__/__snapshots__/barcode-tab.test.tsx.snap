// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`barcode rendering should match with snapshot 1`] = `
[
  <div
    data-testid="input-barcode-text"
  >
    <label
      className="fui-Label ___1yevxz2_yljzsa0 fk6fouc f19n0e5 fkhj508 f1i3iumi"
      htmlFor="fui-r0"
    >
      <b>
        Barcode text
      </b>
    </label>
    <span
      className="fui-Input r1jtohuq ___gxxrxg0_glgvo90 fvcxoqz f1ub3y4t f1l4zc64 f1m52nbi f8vnjqi fz1etlk f1klwx88 f1hc16gm fly5x3f ffemd8d"
    >
      <input
        className="fui-Input__input rvp2gzh"
        id="fui-r0"
        onChange={[Function]}
        placeholder="Type something or select from your document"
        role="textbox"
        type="text"
        value=""
      />
    </span>
  </div>,
  <div
    className="ms-Slider ms-Slider-enabled ms-Slider-row ___15qz5iz_0000000 f18og1rn root-109"
  >
    <label
      className="ms-Label titleLabel-120"
      htmlFor="Slider0"
    >
      Width
    </label>
    <div
      className="ms-Slider-container container-111"
    >
      <div
        aria-disabled={false}
        aria-label="Width"
        aria-valuemax={5}
        aria-valuemin={0.5}
        aria-valuenow={1}
        aria-valuetext="1"
        className="ms-Slider-slideBox ms-Slider-showValue slideBox-112"
        data-is-focusable={true}
        id="Slider0"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
      >
        <div
          className="ms-Slider-line line-114"
        >
          <span
            className="ms-Slider-thumb thumb-113"
            style={
              {
                "left": "11.11111111111111%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "0%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-active activeSection-116"
            style={
              {
                "width": "11.11111111111111%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "88.88888888888889%",
              }
            }
          />
        </div>
      </div>
      <label
        className="ms-Label ms-Slider-value valueLabel-121"
      >
        1
      </label>
    </div>
  </div>,
  <div
    className="ms-Slider ms-Slider-enabled ms-Slider-row ___15qz5iz_0000000 f18og1rn root-109"
  >
    <label
      className="ms-Label titleLabel-120"
      htmlFor="Slider1"
    >
      Height
    </label>
    <div
      className="ms-Slider-container container-111"
    >
      <div
        aria-disabled={false}
        aria-label="Height"
        aria-valuemax={150}
        aria-valuemin={50}
        aria-valuenow={100}
        aria-valuetext="100"
        className="ms-Slider-slideBox ms-Slider-showValue slideBox-112"
        data-is-focusable={true}
        id="Slider1"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
      >
        <div
          className="ms-Slider-line line-114"
        >
          <span
            className="ms-Slider-thumb thumb-113"
            style={
              {
                "left": "50%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "0%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-active activeSection-116"
            style={
              {
                "width": "50%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "50%",
              }
            }
          />
        </div>
      </div>
      <label
        className="ms-Label ms-Slider-value valueLabel-121"
      >
        100
      </label>
    </div>
  </div>,
  <div>
    <button
      className="fui-Button r1alrhcs ___1akj6hk_ih97uj0 ffp7eso f1p3nwhy f11589ue f1q5o8ev f1pdflbu f1phragk f15wkkf3 f1s2uweq fr80ssc f1ukrpxl fecsdlb f1rq72xc fnp9lpt f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1d6v5y2 f1rirnrt f1uu00uk fkvaka8 f1ux7til f9a0qzu f1lkg8j3 fkc42ay fq7113v ff1wgvm fiob0tu f1j6scgf f1x4h75k f4xjyn1 fbgcvur f1ks1yx8 f1o6qegi fcnxywj fmxjhhp f9ddjv3 f17t0x8g f194v5ow f1qgg65p fk7jm04 fhgccpy f32wu9k fu5nqqq f13prjl2 f1czftr5 f1nl83rv f12k37oa fr96u23"
      data-testid="btn-insert-image"
      onClick={[Function]}
      type="button"
    >
      Insert Image
    </button>
  </div>,
]
`;
